apps/settings/js/modules/bluetooth/bluetooth_context.js

```diff
@@ -0,0 +1,809 @@
+     */
+    stopDiscovery: function btc_stopDiscovery() {
+      if ((this.discovering === false) || (this.state !== 'enabled')) {
+        return Promise.reject('same state or Bluetooth is disabled!!');
+      }
+
+      if (!this._defaultAdapter) {
+        return Promise.reject('default adapter is not existed!!');
+      }
+
+      return this._defaultAdapter.stopDiscovery().then(() => {
+        Debug('stopDiscovery(): stopDiscovery successfully :)');
+      }, (reason) => {
+        Debug('stopDiscovery(): stopDiscovery failed: reason = ' + reason);
+        return Promise.reject(reason);
+      });
+    },
+
+    /**
+     * A function to receive BluetoothDiscoveryHandle. And set a function to
+     * handle 'ondevicefound' event.
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @param {Object BluetoothDiscoveryHandle} handle
+     */
+    _setDiscoveryHandler: function btc__setDiscoveryHandler(handle) {
+      Debug('_setDiscoveryHandler(): handle = ' + handle);
+      // make the code base easy to do unit test
+      this._discoveryHandler = handle;
+      this._discoveryHandler.ondevicefound = this._onDeviceFound.bind(this);
+    },
+
+    /**
+     * Handle 'ondevicefound' event to access remote device in list with
+     * observable object.
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @param {Object} evt
+     */
+    _onDeviceFound: function btc__onDeviceFound(evt) {
+      // save device
+      this._saveDevice(evt.device);
+    },
+
+    /**
+     * To distinguish between paired and unpaired.
+     * Then, save/update the device in list corresponding to the state of
+     * property paired.
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @param {Object Observable} device
+     */
+    _saveDevice: function btc__saveDevice(device) {
+      // distinguish the found device is paired or not
+      var operatingDevices =
+        (device.paired) ? this.getPairedDevices() : this.getRemoteDevices();
+
+      // check the device is existed or not in remote/paired devices array
+      var index = operatingDevices.array.findIndex(
+                    this._findDeviceByAddress.bind(this, device.address));
+
+      // If the device is not existed yet, create observable object
+      // for saving this device.
+      if (index === -1) {
+        // create observable BtDevice
+        var observableBtDevice = BtDevice(device);
+        // push device in devices list with observable object
+        operatingDevices.push(observableBtDevice);
+      } else {
+        // The device is existed, no need to do any thing here.
+        // set device in devices list
+        // operatingDevices.set(index, device);
+      }
+    },
+
+    /**
+     * Given address to find out device element from array.
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @param {String} address
+     * @param {Observable BluetoothDevice} btDevice
```

Comment:
Simply `{BluetoothDevice}` as we can learn that `BluetoothDevice` is an `Observable` from the description.